### PR TITLE
Add DSACryptoServiceProvider to Csp assembly

### PR DESCRIPTION
**Pull request by steveharter**
This is the last part of the work for adding DSA per issue #4813  

Note most of the code in CapiHelper.DSA.cs was ported from full desktop c++ code (not managed) using the same style as RSACryptoServiceProvider with loosely-coupled binary reads and writes with BinaryReader, etc. This is not like the code in Algorithms which uses more strongly-typed structs and and helper methods to read or write blobs.

@bartonjs please take a look.

### CODE REVIEW

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+        }
+
+        /// <summary>
+        /// Helper for DSACryptoServiceProvider.ExportParameters()
+        /// </summary>
+        internal static DSAParameters ToDSAParameters(this byte[] cspBlob, bool includePrivateParameters, SafeKeyHandle safeKeyHandle)
+        {
+            try
+            {
+                BinaryReader br = new BinaryReader(new MemoryStream(cspBlob));
```
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: Will add Using statement here for BinaryReader, even though there's not really a handle to anything. The RSA side did not have a Using statement either

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+                        dsaParameters.Y = br.ReadReversed(len);
+
+                    ReadDSSSeed(dsaParameters, br, false);
+
+                    if (includePrivateParameters)
+                    {
+                        // Since DSSPUBKEY key is used for both public and private keys, we got X
+                        // but not Y. To get Y, do another export and ask for public key blob.
+                        byte[] cspPublicBlob = ExportKeyBlob(false, safeKeyHandle);
+                        BinaryReader brPublicBlob = new BinaryReader(new MemoryStream(cspPublicBlob));
```
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: Will add Using statement here for BinaryReader

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+            if (dsaParameters.Seed != null && dsaParameters.Seed.Length != 20)
+                throw NTE_BAD_DATA.ToCryptographicException();
+
+            bool isPrivate = (dsaParameters.X != null && dsaParameters.X.Length != 0);
+
+            uint bitLenP = (uint)dsaParameters.P.Length * 8;
+            uint bitLenJ = dsaParameters.J == null ? 0 : (uint)dsaParameters.J.Length * 8;
+
+            MemoryStream ms = new MemoryStream();
+            BinaryWriter bw = new BinaryWriter(ms);
```
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: Will add Using statement here for BinaryWriter

File: src/System.Security.Cryptography.Csp/src/project.json [hunk 4]
```diff
@@ -1,27 +1,29 @@
...
+        "System.Reflection": "4.3.0-beta-devapi-24512-01",
+        "System.Resources.ResourceManager": "4.3.0-beta-devapi-24512-01",
+        "System.Runtime": "4.3.0-beta-devapi-24512-01",
+        "System.Runtime.Extensions": "4.3.0-beta-devapi-24512-01",
+        "System.Runtime.InteropServices": "4.3.0-beta-devapi-24512-01",
+        "System.Security.Cryptography.Algorithms": "4.3.0-beta-devapi-24512-01",
+        "System.Security.Cryptography.Encoding": "4.3.0-beta-devapi-24512-01",
+        "System.Text.Encoding": "4.3.0-beta-devapi-24512-01",
+        "System.Threading": "4.3.0-beta-devapi-24512-01",
+        "System.Threading.Tasks": "4.3.0-beta-devapi-24512-01"
```
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/project.json]: FYI the new references to Reflection and Threading.Tasks are needed to avoid ambiguous type compile error msgs

File: src/System.Security.Cryptography.Csp/src/System.Security.Cryptography.Csp.csproj [hunk 5]
```diff
@@ -43,11 +45,14 @@
...
   </ItemGroup>
-  <ItemGroup Condition="'$(TargetGroup)' == 'net46'">
+  <ItemGroup Condition="'$(TargetGroup)' == 'net463'">
     <TargetingPackReference Include="mscorlib" />
   </ItemGroup>
   <ItemGroup>
     <None Include="project.json" />
   </ItemGroup>
+  <ItemGroup Condition="'$(IsPartialFacadeAssembly)' != 'true' AND '$(TargetsWindows)' == 'true'">
+    <AdditionalFiles Include="$(ToolsDir)PinvokeAnalyzer_OneCoreApis.txt" />
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System.Security.Cryptography.Csp.csproj]: I made a fix in buildtools to cause this to get picked up correctly, so you shouldn't need it.  And since you don't seem to be adding a netstandard/netcore/netcoreapp TargetGroup you shouldn't have needed it anyways?

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+            return wszUpgrade;
+        }
+
+        /// <summary>
+        /// Helper for DsaCryptoServiceProvider.ImportParameters()
+        /// </summary>
+        internal static byte[] ToKeyBlob(this DSAParameters dsaParameters, int algId)
+        {
+            if (algId != CapiHelper.CALG_DSS_SIGN)
+                throw new PlatformNotSupportedException();
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: If this is the only valid value, why is it a parameter at all?

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+
+            // Y is present for V3 DSA key blobs, Y = g^j mod P
+            if (dsaParameters.Y != null && dsaParameters.Y.Length != dsaParameters.P.Length)
+                throw NTE_BAD_DATA.ToCryptographicException();
+
+            // The seed is always a 20 byte array
+            if (dsaParameters.Seed != null && dsaParameters.Seed.Length != 20)
+                throw NTE_BAD_DATA.ToCryptographicException();
+
+            bool isPrivate = (dsaParameters.X != null && dsaParameters.X.Length != 0);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: The equivalent native check here was `if (X != null && X.Length != DSS_Q_LEN)`, what's the check for 0 length for?

Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: Sorry, I conflated two things.  This is how it checked isPrivate; but you seem to have lost the X length == Q length check.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+                {
+                    dsaParameters.Counter = counter;
+                    dsaParameters.Seed = br.ReadReversed(20);
+                    return;
+                }
+            }
+
+            dsaParameters.Counter = 0;
+            dsaParameters.Seed = null;
+            br.BaseStream.Position += 20; // Advance past seed[20]
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: Maybe make all 3 20s be a method-local const?

File: src/Native/Unix/System.Security.Cryptography.Native/pal_dsa.cpp [hunk 9]
```diff
@@ -126,7 +126,10 @@ extern "C" int32_t CryptoNative_GetDsaParameters(
     *q = dsa->q; *qLength = BN_num_bytes(*q);
     *g = dsa->g; *gLength = BN_num_bytes(*g);
     *y = dsa->pub_key; *yLength = BN_num_bytes(*y);
-    *x = dsa->priv_key; *xLength = BN_num_bytes(*x);
+
+    // dsa->priv_key is optional
+    *x = dsa->priv_key;
+    (*x) == nullptr ? *xLength = 0 : *xLength = BN_num_bytes(*x);
```
Comment (stephentoub) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_dsa.cpp]: Nit: I find this construction hard to follow.  Could it be changed to:
```C++
*xLength = (*x == nullptr) ? 0 : BN_num_bytes(*x);
```
?

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs [hunk 10]
```diff
@@ -0,0 +1,506 @@
...
+
+                    if (current != null)
+                    {
+                        SafeKeyHandle keyHandle = _safeKeyHandle;
+                        _safeKeyHandle = null;
+                        keyHandle?.Dispose();
+                    }
+
+                    _safeProvHandle = value;
+                    current?.Dispose();
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Nit: should this be done in the current != null block, such that you could then just use ```current.Dispose()```?

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs (hunk 6 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: Removed it. The pattern was borrowed from RSA which does the same thing with the comment:
            // The original FCall this helper emulates supports other algId's - however, the only algid we need to support is CALG_RSA_KEYX. We will not
            // port the codepaths dealing with other algid's.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs (hunk 7 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: Good catch; thanks for looking at the native code. I added the check for == DSS_Q_LEN if the length is greater than 0. I believe the is a x mod q, so that makes sense, even though xbitlen is still part of the structure

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs (hunk 8 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: I refactored it slightly to combine the first 2 cases, but left the last += 20 like that for efficiency

File: src/Native/Unix/System.Security.Cryptography.Native/pal_dsa.cpp (hunk 9 above)
Comment (steveharter) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_dsa.cpp]: Sure. I made that change in nano where I don't edit very well :)

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs (hunk 10 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Changed it. Also changed the RSA where this was copied from. The keyHandle?. could probably be just keyHandle as it appears a lock (_parameters) is used when it is created, but I left that as-is.

File: src/System.Security.Cryptography.Csp/src/System.Security.Cryptography.Csp.csproj (hunk 5 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System.Security.Cryptography.Csp.csproj]: Removed it. At the time, I needed it as I was building 1.6 and the tools change wasn't there.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+        internal static DSAParameters ToDSAParameters(this byte[] cspBlob, bool includePrivateParameters, SafeKeyHandle safeKeyHandle)
+        {
+            try
+            {
+                using (var ms = new MemoryStream(cspBlob))
+                using (var br = new BinaryReader(ms))
+                {
+                    byte bType;
+                    byte bVersion;
+                    ReadKeyBlobHeader(br, out bType, out bVersion);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: Why are you outing bType to this method, you don't ever seem to read it.

In theory bType is redundant to includePrivateParameters, so you could either assert they stayed aligned, or remove includePrivateParameters from the signature and build it off of bType.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs (hunk 10 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: If something read SafeProvHandle (into existence) but never read SafeKeyHandle (into existence) and then changed SafeProvHandle, _safeKeyHandle would be null.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs [hunk 12]
```diff
@@ -0,0 +1,506 @@
...
+
+                    _safeKeyHandle = value;
+                    current?.Dispose();
+                }
+            }
+        }
+
+        /// <summary>
+        /// CspKeyContainerInfo property
+        /// </summary>
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: This doc comment doesn't really add value...

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs [hunk 13]
```diff
@@ -0,0 +1,506 @@
...
+                SafeKeyHandle localHandle = SafeKeyHandle;
+                Debug.Assert(localHandle != null);
+
+                return new CspKeyContainerInfo(_parameters, _randomKeyContainer);
+            }
+        }
+
+        /// <summary>
+        /// KeySize property
+        /// </summary>
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Superfluous doc comment.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs [hunk 14]
```diff
@@ -0,0 +1,506 @@
...
+        {
+            get
+            {
+                return new[] { new KeySizes(512, 1024, 64) }; // per FIPS 186-2
+            }
+        }
+
+        /// <summary>
+        /// Persisted key in CSP 
+        /// </summary>
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Superfluous doc comment.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs [hunk 15]
```diff
@@ -0,0 +1,506 @@
...
+                {
+                    return; // Do nothing
+                }
+                CapiHelper.SetPersistKeyInCsp(SafeProvHandle, value);
+            }
+        }
+
+        /// <summary>
+        /// Gets the information of key if it is a public key
+        /// </summary>
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Superfluous doc comment.  (It actually only gets whether or not we believe it's a public key.  And I think it only identifies that when importing from bytes, so a named public-only key probably returns false)

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs [hunk 16]
```diff
@@ -0,0 +1,506 @@
...
+            get
+            {
+                byte[] publicKey = CapiHelper.GetKeyParameter(SafeKeyHandle, Constants.CLR_PUBLICKEYONLY);
+                return (publicKey[0] == 1);
+            }
+        }
+
+        /// <summary>
+        /// MachineKey store properties
+        /// </summary>
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Superfluous doc comment.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs [hunk 17]
```diff
@@ -0,0 +1,506 @@
...
+            }
+            set
+            {
+                s_useMachineKeyStore = (value ? CspProviderFlags.UseMachineKeyStore : 0);
+            }
+        }
+
+        /// <summary>
+        /// Dispose the key handles 
+        /// </summary>
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Superfluous doc comment.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs [hunk 18]
```diff
@@ -0,0 +1,506 @@
...
+        {
+            SafeProvHandle safeProvHandle;
+            CapiHelper.AcquireCsp(new CspParameters(CapiHelper.DefaultDssProviderType), out safeProvHandle);
+            return safeProvHandle;
+        }
+
+        /// <summary>
+        /// Imports a blob that represents DSA key information
+        /// </summary>
+        /// <param name="keyBlob"></param>
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Some of these comments look like they're being scraped from MSDN, but they're usually missing the param entries; and when present (like this one) the param isn't properly documented.  What's up with the comments?

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs [hunk 19]
```diff
@@ -0,0 +1,506 @@
...
+            return VerifyHash(rgbHash, null, rgbSignature);
+        }
+
+        protected override byte[] HashData(byte[] data, int offset, int count, HashAlgorithmName hashAlgorithm)
+        {
+            // we're sealed and the base should have checked this before calling us
+            Debug.Assert(data != null);
+            Debug.Assert(offset >= 0 && offset <= data.Length);
+            Debug.Assert(count >= 0 && count <= data.Length - offset);
+            Debug.Assert(!String.IsNullOrEmpty(hashAlgorithm.Name));
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Style: String => string (I see at least one other hit)

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs [hunk 20]
```diff
@@ -0,0 +1,506 @@
...
+
+            return true;
+        }
+
+        /// <summary>
+        /// Verifies the signature of a hash value.
+        /// </summary>
+        private bool VerifyHash(byte[] rgbHash, int calgHash, byte[] rgbSignature)
+        {
+            return CapiHelper.VerifySign(
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Why is this inlined for sign, but a different method for verify?

File: src/System.Security.Cryptography.Csp/tests/DSACryptoServiceProviderTests.cs [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+                }
+            }
+        }
+
+        [Fact]
+        public static void NamedKey_AlternateProvider()
+        {
+            const int KeySize = 512;
+
+            CspParameters cspParameters = new CspParameters(PROV_DSS_DH)
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/tests/DSACryptoServiceProviderTests.cs]: This looks to be the same provider as the previous test... what's the actual difference?

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs (hunk 11 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.DSA.cs]: I will remove the out parameter. The variable was useful for debugging sanity check, and I also had an Assert at one point which I removed for concern over compatibility. It is not verified or used otherwise for compatibility (as done on RSA side as well). Here's the comment on the RSA side:
// BLOBHEADER.bType: Expected to be 0x6 (PUBLICKEYBLOB) or 0x7 (PRIVATEKEYBLOB), though there's no check for backward compat reasons.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs (hunk 12 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: I agree. Those documents were copy-paste from RSA side... I'll change to the msdn info

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs (hunk 15 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Changed to msdn: "Gets a value that indicates whether the DSACryptoServiceProvider object contains only a public key."

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs (hunk 18 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Actually not sure what we should be doing here regarding best practice. Most originated from RSA side (which were superfluous), but I changed some of the 'important' ones to match msdn. Now I plan to remove the comments if they are overrides of base class or simple properties (redundant wording) and adding any missing msdn comments.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs (hunk 20 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/DSACryptoServiceProvider.cs]: Just Inlined it. Remnant of RSA side

File: src/System.Security.Cryptography.Csp/tests/DSACryptoServiceProviderTests.cs (hunk 21 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/tests/DSACryptoServiceProviderTests.cs]: That was missed clean-up on my side. Remnant from RSA which has two supported providers. Removed the test

**Comment by bartonjs** [comment]
Please squash :)
